<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین ۲<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 style="color:rgb(90, 255, 184); font-size: 20px;">Logistic Regression and Naive Bayes</h2>
<p align="center" style="color: #666; font-size: 16px;">فرهاد نصری - علیرضا زمانی</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">farhadnasri999@gmail.com - shigzv@gmail.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: سید مانی میرشعبانی</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: ۸۱۰۸۰۱۰۸۰</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال: {{تاریخ_ارسال}}</p>
</div>
</div>


<div dir="rtl" style="border: 2px solid #f0ad4e; background-color: #fffbe6; border-radius: 8px; padding: 20px; margin: 20px auto; width: 90%; text-align: justify; font-family: Vazir, sans-serif; color: #333; box-shadow: 0 2px 4px rgba(0,0,0,0.1);">
    <p style="font-weight: bold; margin-bottom: 15px; color: #d9534f; font-size: 1.1em;">
        ⚠️ توضیح در خصوص استفاده از دستیار هوش مصنوعی:
    </p>
    <p style="line-height: 1.8;">
        این تمرین با کمک دستیار هوش مصنوعی <strong>Gemini</strong> انجام شده است و مشروح مکالمه از طریق این لینک قابل مشاهده است: 
        <a href="https://gemini.google.com/share/e5b66b139c5d" target="_blank" style="color: #0275d8; text-decoration: none; font-weight: bold;">(لینک)</a>. 
        <br><br>
        در تمامی مراحل، حالت <strong>«Guided Learning»</strong> فعال بوده و صرفاً جهت بررسی خطاها (Error Checking) مورد استفاده قرار گرفته است. 
        همچنین از این ابزار برای ویرایش و زیباسازی قالب‌بندی (Markup) استفاده شده است؛ بدین صورت که متن اصلی مارک‌آپ‌ها توسط خودم به زبان انگلیسی نگارش شده و سپس جهت ساختاردهی صحیح مارک‌داون و نمایش مناسب، توسط هوش مصنوعی ویرایش شده‌اند.
    </p>
</div>


<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<li><b>سوال اول - <span dir="ltr">Logistic Regression and Naive Bayes (from scratch)</span> (60)</b></li>
<ul>
<li>بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</li>
<li>بخش دوم: جداسازی داده‌های آموزش و آزمایش</li>
<li>بخش سوم: پیاده‌سازی معیار‌های ارزیابی</li>
<li>بخش چهارم: پیاده‌سازی Logistic Regression</li>
<li>بخش پنجم: پیاده‌سازی Naive Bayes</li>
<li>بخش ششم: تحلیل نتایج</li>
</ul>
<li><b>سوال دوم - <span dir="ltr">Logistic Regression and Naive Bayes (e.g. 
sklearn)</span> (40)</b></li>
<ul>
<li>بخش اول: استخراج ویژگی‌های ساختاری</li>
<li>بخش دوم: استخراج ویژگی‌های آماری و محتوایی</li>
<li>بخش سوم: استخراج ویژگی‌های دلخواه</li>
<li>بخش چهارم: ترکیب همه ویژگی‌های استخراج‌شده</li>
</ul>
</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
در سوال دو مجاز به استفاده از کتابخانه‌های آماده برای مدل‌ها و متریک‌ها هستید ولی در سوال یک، همان‌طور که در متن سوال هم ذکر شده، باید از پایه پیاده‌سازی کنید.
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال اول: پیاده‌سازی Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این سوال، شما باید این دو مدل پایه‌ای برای طبقه‌بندی متون را از صفر (بدون استفاده از کتابخانه‌های آماده) پیاده‌سازی کنید. هم‌چنین نیاز است، پیش‌پردازش داده‌ها و ارزیابی عملکرد مدل‌ها را نیز انجام دهید.




## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: پیش پردازش داده‌های متنی و استفاده از Bag of Words</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در گام پیش‌پردازش، هدف ما آماده‌سازی داده‌های متنی برای استفاده در مدل‌های یادگیری ماشین است. از آن‌جا که مدل‌ها با داده‌های عددی کار می‌کنند، باید متن خام را به شکلی عددی تبدیل کنیم تا بتوانند الگوهای موجود در واژه‌ها را بیاموزند. در این بخش، قصد داریم با تبدیل متون به نمایش Bag of Words، هر ایمیل را بر اساس تعداد تکرار واژه‌هایش به یک بردار عددی تبدیل کنیم.
<br>
دیتاست emails_1.csv شامل دو ستون است:
<br>
text (متن ایمیل)
<br>
status (برچسب، مشخص‌کننده‌ی spam یا ham بودن ایمیل)
<br>
شما باید مراحل زیر را انجام دهید:
<br>
۱. ستون status را به مقادیر عددی تبدیل کنید (spam → 1 و ham → 0).
<br>
۲. متن‌ها را پاک‌سازی کنید: تمام کاراکترهای غیر الفبایی (اعداد، علائم و غیره) را حذف و همه‌ی حروف را کوچک (lowercase) کنید.
<br>
۳. با استفاده از روش Bag of Words، فراوانی کلمات را محاسبه کنید و فقط ۱۵ کلمه‌ی پرتکرار را نگه دارید.
<br>
💡 نکته: برای این بخش می‌توانید از CountVectorizer در کتابخانه‌ی sklearn استفاده کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
یک دیتافریم جدید که شامل ۱۵ ویژگی + ستون status است. (تمام ۱۰ ردیف این دیتافریم را نمایش دهید.)
</p>

For this section we will first take a quick look at the dataset.

In [15]:
import pandas as pd
emails1_dataset = pd.read_csv('./datasets/q1/emails_1.csv')

emails1_dataset.head()

,text,status
0,🎉 Congratulations! You have won a FREE ticket ...,spam
1,Reminder: The project meeting will be held tom...,ham
2,Exclusive offer just for YOU! Get a free-trial...,spam
3,Hey team - the report looks great; please revi...,ham
4,You’ve been selected for a *FREE* gift card! C...,spam


Then we will convert the status section into binary numbers.

In [16]:
status_mapping = {'ham': 0, 'spam': 1}
emails1_dataset['status'] = emails1_dataset['status'].map(status_mapping)

Next we will remove all non alphabetic characters form the emails and convert all the characters to lowercase.

In [17]:
def preprocess_email(text):
    cleaned_text = []
    for char in text:
        if char.isalpha() or char.isspace():
            cleaned_text.append(char.lower())
    cleaned_text = ''.join(cleaned_text)
    return ' '.join(cleaned_text.split())

emails1_dataset['text'] = emails1_dataset['text'].apply(preprocess_email)

Next we shall use the Bag of Words method to count the frequency of each word and keep the 15 most frequent words

In [18]:
def calculate_word_frequencies(corpus):
    word_freq = {}
    for text in corpus:
        words = text.split()
        for word in words:
            if word in word_freq:
                word_freq[word] += 1
            else:
                word_freq[word] = 1
    
    return word_freq

word_frequencies = calculate_word_frequencies(emails1_dataset['text'])
sorted_word_freq = dict(sorted(word_frequencies.items(), key=lambda item: item[1], reverse=True)[:15])
print(sorted_word_freq)

{'your': 10, 'free': 6, 'for': 6, 'the': 5, 'to': 4, 'you': 3, 'a': 3, 'our': 3, 'claim': 3, 'project': 3, 'please': 3, 'and': 3, 'get': 3, 'ticket': 2, 'it': 2}


Now that we have found the most frequent words, we shall create a feature for each of them in our dataset.

In [19]:
def word_count(text, word):
    words = text.split()
    return words.count(word)

for word in sorted_word_freq.keys():
    emails1_dataset[word] = emails1_dataset['text'].apply(lambda x: word_count(x, word))

emails1_dataset.drop(columns=['text'], inplace=True)
emails1_dataset.head()

,status,your,free,for,the,to,you,a,our,claim,project,please,and,get,ticket,it
0,1,1,2,0,0,1,1,1,1,1,0,0,0,0,2,1
1,0,1,0,0,1,0,0,0,0,0,2,1,1,0,0,0
2,1,0,1,2,0,0,1,1,1,0,0,0,0,1,0,0
3,0,1,0,0,1,0,0,0,0,0,0,1,1,0,0,1
4,1,2,1,1,0,0,0,1,0,1,0,0,0,0,0,0


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
از این مرحله به بعد، دیگر با فایل emails_1.csv و دیتافریم حاصل از آن کاری نخواهیم داشت. برای سهولت کار، فایل emails_2.csv در اختیار شما قرار گرفته است. این فایل، نتیجه‌ی همان فرایند پیش‌پردازش بر روی ۵۱۷۲ ایمیل واقعی است. به بیان ساده‌تر، در این مجموعه داده، هر ردیف نمایانگر یک ایمیل و هر ستون نشان‌دهنده‌ی فراوانی یکی از ۳۰۰۰ کلمه‌ی پرتکرار در کل داده‌ها است.
در ادامه، مدل‌هایی که پیاده‌سازی می‌کنید باید روی این مجموعه داده، آموزش داده شوند. بنابراین، در این بخش فایل emails_2.csv را بخوانید و محتوای آن را نمایش دهید تا با ساختار داده آشنا شوید.
</p>



<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
پنج ردیف اول دیتافریم مذکور را نمایش دهید.
</p>

In [20]:
emails2_dataset = pd.read_csv('./datasets/q1/emails_2.csv')
emails2_dataset.head()

,Email No.,the,to,ect,and,for,of,a,you,hou,...,connevey,jay,valued,lay,infrastructure,military,allowing,ff,dry,Prediction
0,Email 1,0,0,1,0,0,0,2,0,0,...,0,0,0,0,0,0,0,0,0,0
1,Email 2,8,13,24,6,6,2,102,1,27,...,0,0,0,0,0,0,0,1,0,0
2,Email 3,0,0,1,0,0,0,8,0,0,...,0,0,0,0,0,0,0,0,0,0
3,Email 4,0,5,22,0,5,1,51,2,10,...,0,0,0,0,0,0,0,0,0,0
4,Email 5,7,6,17,1,5,2,57,0,9,...,0,0,0,0,0,0,0,1,0,0


In [21]:
emails2_dataset.dtypes

Email No.     object
the            int64
to             int64
ect            int64
and            int64
               ...  
military       int64
allowing       int64
ff             int64
dry            int64
Prediction     int64
Length: 3002, dtype: object

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: جداسازی داده‌های آموزش و آزمایش</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
داده‌ها را با نسبت ۸۰ درصد برای آموزش و ۲۰ درصد برای آزمایش تقسیم کنید. در انتها تعداد نمونه‌ها را چاپ کنید.
</p>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
X_train, X_test, y_train, y_test<br>
تعداد نمونه‌های X_train و X_test 
</p>

For this section I assumed it's fine to use sklearn since it was in of the hints before and the task of splitting the data manually isn't really that difficult!

We divide the data into 80% training and 20% test, we then further divide the training data to 80% for training the model and 20% for validation.

In [22]:
!pip install scikit-learn

In [23]:
from sklearn.model_selection import train_test_split

random_state = 24
X_train, X_test, y_train, y_test = train_test_split(
    emails2_dataset.drop(columns=['Prediction', 'Email No.']),
    emails2_dataset['Prediction'],
    test_size=0.2,
    random_state=random_state
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=random_state
)

print("Training set size:", X_train.shape)
print("Validation set size:", X_val.shape)
print("Test set size:", X_test.shape)

Training set size: (3309, 3000)
Validation set size: (828, 3000)
Test set size: (1035, 3000)


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: پیاده‌سازی معیار‌های ارزیابی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
پیش از آن‌که وارد مرحله‌ی پیاده‌سازی و آموزش مدل‌های طبقه‌بندی شویم، لازم است معیارهایی برای سنجش عملکرد آن‌ها تعریف کنیم. این معیارها به ما کمک می‌کنند تا بفهمیم مدل تا چه اندازه در تشخیص درست نمونه‌های مثبت و منفی موفق عمل کرده است.
<br>
در این بخش، توابع زیر را برای حالت دودویی (binary classification) پیاده‌سازی کنید:
<br>
accuracy(y_true, y_pred) — نسبت پیش‌بینی‌های درست به کل نمونه‌ها
<br>
precision(y_true, y_pred) — درصد پیش‌بینی‌های مثبت که واقعاً مثبت بوده‌اند
<br>
recall(y_true, y_pred) — درصد نمونه‌های مثبت واقعی که مدل آن‌ها را درست شناسایی کرده است
<br>
f1_score(y_true, y_pred) — میانگین هارمونیک بین precision و recall، برای ایجاد توازن میان آن دو
<br>
هر تابع باید مقدار عددی متناظر با معیار مورد نظر را برگرداند.
<br>
 به توابع بالا دو ورودی زیر را بدهید و خروجی بگیرید:    y_true = [0, 1, 1, 0, 1] ---- y_pred = [0, 1, 0, 0, 1]
<br>
💡 نکته: برای این بخش می‌توانید از  کتابخانه‌ی numpy استفاده کنید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
مثال:
<br>
y_true = [0, 1, 1, 0, 1]
<br>
y_pred = [0, 1, 0, 0, 1]
<br>
print(accuracy(y_true, y_pred))  # 0.80
<br>
print(precision(y_true, y_pred)) # 1.00
<br>
print(recall(y_true, y_pred))    # 0.66
<br>
print(f1_score(y_true, y_pred))  # 0.80
<br>
</p>

In this section we will create a class called Metrics and implement all of these metrics as its methods.

In [24]:
class Metrics:
    def __init__(self):
        pass
    
    @staticmethod
    def accuracy(y_true, y_pred):
        correct_predictions = sum(y_true == y_pred)
        return correct_predictions / len(y_true)
    @staticmethod
    def precision(y_true, y_pred):
        true_positives = sum((y_true == 1) & (y_pred == 1))
        predicted_positives = sum(y_pred == 1)
        return true_positives / predicted_positives if predicted_positives > 0 else 0
    @staticmethod
    def recall(y_true, y_pred):
        true_positives = sum((y_true == 1) & (y_pred == 1))
        actual_positives = sum(y_true == 1)
        return true_positives / actual_positives if actual_positives > 0 else 0
    @staticmethod
    def f1_score(y_true, y_pred):
        prec = Metrics.precision(y_true, y_pred)
        rec = Metrics.recall(y_true, y_pred)
        return 2 * (prec * rec) / (prec + rec) if (prec + rec) > 0 else 0   

Now we test it on a simple data.

In [25]:
import numpy as np

y_true = np.array([0, 1, 1, 0, 1])
y_pred = np.array([0, 1, 0, 0, 1])

print("Accuracy:", Metrics.accuracy(y_true, y_pred))
print("Precision:", Metrics.precision(y_true, y_pred))
print("Recall:", Metrics.recall(y_true, y_pred))
print("F1 Score:", Metrics.f1_score(y_true, y_pred))

Accuracy: 0.8
Precision: 1.0
Recall: 0.6666666666666666
F1 Score: 0.8


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: پیاده‌سازی Logistic Regression</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Logistic Regression را به طور کامل پیاده سازی کنید.سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به متریک های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
💡استفاده از هایپر پارامتر‌های مناسب مانند نرخ یادگیری و تعداد دوره‌های آموزش بر عهده خودتان است.
<br>
💡میتوانید قبل از آموزش داده ها را نرمال سازی کنید.
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [26]:
class logistic_regression:
    def __init__(self, learning_rate=0.01, num_iterations=1000):
        self.learning_rate = learning_rate
        self.num_iterations = num_iterations
        self.weights = None
        self.bias = None

    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))
    
    def fit(self, X, y):
        num_samples, num_features = X.shape
        self.weights = np.zeros(num_features)
        self.bias = 0

        for _ in range(self.num_iterations):
            linear_model = np.dot(X, self.weights) + self.bias
            y_predicted = self.sigmoid(linear_model)

            dw = (1 / num_samples) * np.dot(X.T, (y_predicted - y))
            db = (1 / num_samples) * np.sum(y_predicted - y)

            self.weights -= self.learning_rate * dw
            self.bias -= self.learning_rate * db

    def predict(self, X):
        linear_model = np.dot(X, self.weights) + self.bias
        y_predicted = self.sigmoid(linear_model)
        y_predicted_cls = [1 if i > 0.5 else 0 for i in y_predicted]
        return np.array(y_predicted_cls)
    
    def predict_proba(self, X):
        linear_model = np.dot(X, self.weights) + self.bias
        y_predicted = self.sigmoid(linear_model)
        return y_predicted

Before training the model on the training data, we will first define a normalizer.

In [27]:
class normalizer:
    def __init__(self):
        self.min = None
        self.range = None
    
    def normalize(self, X):
        if self.min is None:
            self.min = X.min(axis=0)
        if self.range is None:
            self.range = X.max(axis=0) - X.min(axis=0)
            
        X_normalized = (X - self.min) / (self.range + 1e-9)
        return X_normalized

Now we will normalize the data and train a model on it!

In [28]:
normalizer_instance = normalizer()
X_train_normalized = normalizer_instance.normalize(X_train)
X_val_normalized = normalizer_instance.normalize(X_val)
X_test_normalized = normalizer_instance.normalize(X_test)

model = logistic_regression(learning_rate=0.01, num_iterations=10000)
model.fit(X_train_normalized.values, y_train.values)

y_train_pred = model.predict(X_train_normalized.values)
print("Training Accuracy:", Metrics.accuracy(y_train.values, y_train_pred))
print("Training Precision:", Metrics.precision(y_train.values, y_train_pred))
print("Training Recall:", Metrics.recall(y_train.values, y_train_pred))
print("Training F1 Score:", Metrics.f1_score(y_train.values, y_train_pred))

y_val_pred = model.predict(X_val_normalized.values)
print("Validation Accuracy:", Metrics.accuracy(y_val.values, y_val_pred))
print("Validation Precision:", Metrics.precision(y_val.values, y_val_pred))
print("Validation Recall:", Metrics.recall(y_val.values, y_val_pred))
print("Validation F1 Score:", Metrics.f1_score(y_val.values, y_val_pred))

Training Accuracy: 0.8455726805681475
Training Precision: 0.9798657718120806
Training Recall: 0.46595744680851064
Training F1 Score: 0.631578947368421
Validation Accuracy: 0.8357487922705314
Validation Precision: 0.9541984732824428
Validation Recall: 0.49019607843137253
Validation F1 Score: 0.6476683937823834


We can see that the results are not ideal. While the precision is quite high, recall and f1 score are not ideal.

I suspect this is due to the data imbalance, so I will be doing some tests to verify and potential solve the issue.

In [29]:
print(y_train.value_counts())

Prediction
0    2369
1     940
Name: count, dtype: int64


In [30]:
y_train_proba = model.predict_proba(X_train_normalized.values)
y_train_pr = pd.Series(y_train_proba).describe()
print(y_train_pr)

count    3309.000000
mean        0.282370
std         0.207814
min         0.000002
25%         0.148800
50%         0.229765
75%         0.357462
max         0.996222
dtype: float64


As we can see the data is imbalanced so we will try to overcome this by sampling equally from classes and training a new model

In [31]:
def balanced_sampling(X, y):
    df = X.copy()
    df['TST'] = y
    min_count = df['TST'].value_counts().min()
    
    balanced_df = pd.DataFrame()
    for label in df['TST'].unique():
        sampled_df = df[df['TST'] == label].sample(min_count, random_state=random_state)
        balanced_df = pd.concat([balanced_df, sampled_df])
    
    balanced_X = balanced_df.drop(columns=['TST'])
    balanced_y = balanced_df['TST']
    
    return balanced_X, balanced_y

X_train_balanced, y_train_balanced = balanced_sampling(X_train, y_train)
print(y_train_balanced.value_counts())

TST
0    940
1    940
Name: count, dtype: int64


In [32]:
normalizer_balanced = normalizer()
X_train_balanced = normalizer_balanced.normalize(X_train_balanced)
X_train_normalized2 = normalizer_balanced.normalize(X_train)
X_val_normalized2 = normalizer_balanced.normalize(X_val)
X_test_normalized2 = normalizer_balanced.normalize(X_test)

model_balanced = logistic_regression(learning_rate=0.01, num_iterations=10000)
model_balanced.fit(X_train_balanced.values, y_train_balanced.values)

y_train_pred = model_balanced.predict(X_train_normalized2.values)
print("Training Accuracy (Balanced):", Metrics.accuracy(y_train.values, y_train_pred))
print("Training Precision (Balanced):", Metrics.precision(y_train.values, y_train_pred))
print("Training Recall (Balanced):", Metrics.recall(y_train.values, y_train_pred))
print("Training F1 Score (Balanced):", Metrics.f1_score(y_train.values, y_train_pred))

y_val_pred = model_balanced.predict(X_val_normalized2.values)
print("Validation Accuracy (Balanced):", Metrics.accuracy(y_val.values, y_val_pred))
print("Validation Precision (Balanced):", Metrics.precision(y_val.values, y_val_pred))
print("Validation Recall (Balanced):", Metrics.recall(y_val.values, y_val_pred))
print("Validation F1 Score (Balanced):", Metrics.f1_score(y_val.values, y_val_pred))

Training Accuracy (Balanced): 0.9552734965246298
Training Precision (Balanced): 0.8687150837988827
Training Recall (Balanced): 0.9925531914893617
Training F1 Score (Balanced): 0.926514399205561
Validation Accuracy (Balanced): 0.9444444444444444
Validation Precision (Balanced): 0.8542372881355932
Validation Recall (Balanced): 0.9882352941176471
Validation F1 Score (Balanced): 0.9163636363636364


We can see that the model has performed significantly better on both training and validation data! 

Now let's finally check both models on the test data.

In [33]:
y_test_pred1 = model.predict(X_test_normalized.values)
y_test_pred2 = model_balanced.predict(X_test_normalized2.values)

print("Test Set Performance of Original Model:")
print("Test Accuracy:", Metrics.accuracy(y_test.values, y_test_pred1))
print("Test Precision:", Metrics.precision(y_test.values, y_test_pred1))
print("Test Recall:", Metrics.recall(y_test.values, y_test_pred1))
print("Test F1 Score:", Metrics.f1_score(y_test.values, y_test_pred1))
print("\nTest Set Performance of Balanced Model:")
print("Test Accuracy:", Metrics.accuracy(y_test.values, y_test_pred2))
print("Test Precision:", Metrics.precision(y_test.values, y_test_pred2))
print("Test Recall:", Metrics.recall(y_test.values, y_test_pred2))
print("Test F1 Score:", Metrics.f1_score(y_test.values, y_test_pred2))

Test Set Performance of Original Model:
Test Accuracy: 0.808695652173913
Test Precision: 0.9495798319327731
Test Recall: 0.3704918032786885
Test F1 Score: 0.5330188679245282

Test Set Performance of Balanced Model:
Test Accuracy: 0.9458937198067633
Test Precision: 0.8487394957983193
Test Recall: 0.9934426229508196
Test F1 Score: 0.9154078549848943


As we can see the balanced model has a significantly better performance on the test data, reaching a f1-score of 0.91 and a Recall of 0.99! The Precision is higher on the original model but Precision alone isn't enough for a good model if Recall is too low.

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: پیاده‌سازی Naive Bayes</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
کلاس Multinomial Naive Bayes را به طور کامل پیاده سازی کنید. سپس مدل را بر روی داده‌های train آموزش دهید و در انتها دقت مدل را باتوجه به معیار های تعریف شده بر روی داده‌های test گزارش کنید.
<br>
- چرا از مدل‌های دیگر مانند Gaussian Naive Bayes استفاده نکردیم؟ آیا از مدل‌های دیگر Naive Bayes می‌توان استفاده کرد؟
<br>
💡مجاز به استفاده از کتابخانه اماده نیستید.
<br>
💡برای آشنایی با Naive Bayes می‌توانید به Appendix کتاب jurafsky مراجعه کنید.
</p>

<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Understanding Naive Bayes</b><br>
Unlike traditional algorithms (like Logistic Regression) that learn weights iteratively, Naive Bayes takes a <b>probabilistic approach</b>. It scans the dataset to calculate the frequency of words and determines the probability of a new email belonging to a class using <b>Bayes' Theorem</b>:

$$P(Spam | Word) = \frac{P(Word | Spam) \cdot P(Spam)}{P(Word)}$$

It is called "Naive" because it assumes <b>conditional independence</b>: it treats every word as if it has no relation to any other word. In reality, this is often false (e.g., "free" and "money" often appear together), but the model still performs remarkably well in practice.
<br><br>

<b>2. Why not Gaussian Naive Bayes?</b><br>
We did not use Gaussian NB because it assumes features follow a <b>Normal (Gaussian) Distribution</b> (the bell curve).
<br>
Our dataset consists of <b>discrete word counts</b> (0, 1, 2...), which creates a very <b>sparse</b> matrix with many zeros. Fitting a continuous Normal distribution to this sparse, discrete data leads to poor assumptions about the probability density, resulting in lower accuracy.
<br><br>

<b>3. Alternative Models</b><br>
The <b>Multinomial Naive Bayes</b> is the standard choice here because it explicitly models discrete frequency counts (multinomial distribution). However, other variations exist:
<ul>
    <li><b>Bernoulli Naive Bayes:</b> Useful if we only cared about the <i>presence</i> or <i>absence</i> of a word (binary features). Since the frequency of specific words matters in our case, this is less ideal than Multinomial.</li>
    <li><b>Complement Naive Bayes:</b> A variation of Multinomial NB that is particularly robust for <b>imbalanced datasets</b>. Instead of calculating the probability of belonging to a class, it calculates the probability of <i>not</i> belonging to all other classes. Given our imbalanced data, this could be a strong candidate.</li>
</ul>
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
معیار‌های پیاده سازی شده را بر روی داده‌های test خروجی بگیرید و نتایج را چاپ کنید
</p>

In [34]:
class multinomial_naive_bayes:
    def __init__(self):
        self.class_priors = {}
        self.word_likelihoods = {}
        self.vocab_size = 0

    def fit(self, X, y):
        num_samples, num_features = X.shape
        self.vocab_size = num_features
        class_counts = y.value_counts().to_dict()

        for cls in class_counts.keys():
            X_cls = X[y == cls]
            total_words_in_class = X_cls.sum().sum()
            self.class_priors[cls] = class_counts[cls] / num_samples
            self.word_likelihoods[cls] = (X_cls.sum() + 1) / (total_words_in_class + self.vocab_size)
    
    def predict(self, X):
        y_pred = []
        for _, x in X.iterrows():
            class_scores = {}
            for cls in self.class_priors.keys():
                log_prior = np.log(self.class_priors[cls])
                log_likelihood = np.sum(x * np.log(self.word_likelihoods[cls]))
                class_scores[cls] = log_prior + log_likelihood
            predicted_class = max(class_scores, key=class_scores.get)
            y_pred.append(predicted_class)
        return np.array(y_pred)

In [35]:
bayes_model = multinomial_naive_bayes()
bayes_model.fit(X_train, y_train)
y_train_pred_bayes = bayes_model.predict(X_train)
print("Naive Bayes Training Accuracy:", Metrics.accuracy(y_train.values, y_train_pred_bayes))
print("Naive Bayes Training Precision:", Metrics.precision(y_train.values, y_train_pred_bayes))
print("Naive Bayes Training Recall:", Metrics.recall(y_train.values, y_train_pred_bayes))
print("Naive Bayes Training F1 Score:", Metrics.f1_score(y_train.values, y_train_pred_bayes))

y_val_pred_bayes = bayes_model.predict(X_val)
print("Naive Bayes Validation Accuracy:", Metrics.accuracy(y_val.values, y_val_pred_bayes))
print("Naive Bayes Validation Precision:", Metrics.precision(y_val.values, y_val_pred_bayes))
print("Naive Bayes Validation Recall:", Metrics.recall(y_val.values, y_val_pred_bayes))
print("Naive Bayes Validation F1 Score:", Metrics.f1_score(y_val.values, y_val_pred_bayes))

Naive Bayes Training Accuracy: 0.9510426110607434
Naive Bayes Training Precision: 0.8874501992031872
Naive Bayes Training Recall: 0.9478723404255319
Naive Bayes Training F1 Score: 0.9166666666666666
Naive Bayes Validation Accuracy: 0.9347826086956522
Naive Bayes Validation Precision: 0.8628158844765343
Naive Bayes Validation Recall: 0.9372549019607843
Naive Bayes Validation F1 Score: 0.8984962406015038


As we can see without any modification or balancing the data, the model already performs quite well.

In [36]:
y_test_pred_bayes = bayes_model.predict(X_test)
print("Naive Bayes Test Accuracy:", Metrics.accuracy(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test Precision:", Metrics.precision(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test Recall:", Metrics.recall(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test F1 Score:", Metrics.f1_score(y_test.values, y_test_pred_bayes))

Naive Bayes Test Accuracy: 0.9497584541062802
Naive Bayes Test Precision: 0.8892307692307693
Naive Bayes Test Recall: 0.9475409836065574
Naive Bayes Test F1 Score: 0.9174603174603175


## <div style="text-align: center; direction: rtl; font-family: Vazir;"> بخش ششم: تحلیل نتایج</div>

<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتایج دو مدل پیاده سازی شده را با یکدیگر مقایسه کنید و تحلیلی بر نتایج داشته باشید.<br>
    
</p>

In [37]:
y_test_pred2 = model_balanced.predict(X_test_normalized2.values)

print("\nTest Set Performance of Balanced Model:")
print("Test Accuracy:", Metrics.accuracy(y_test.values, y_test_pred2))
print("Test Precision:", Metrics.precision(y_test.values, y_test_pred2))
print("Test Recall:", Metrics.recall(y_test.values, y_test_pred2))
print("Test F1 Score:", Metrics.f1_score(y_test.values, y_test_pred2))


Test Set Performance of Balanced Model:
Test Accuracy: 0.9458937198067633
Test Precision: 0.8487394957983193
Test Recall: 0.9934426229508196
Test F1 Score: 0.9154078549848943


In [38]:
y_test_pred_bayes = bayes_model.predict(X_test)
print("Naive Bayes Test Accuracy:", Metrics.accuracy(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test Precision:", Metrics.precision(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test Recall:", Metrics.recall(y_test.values, y_test_pred_bayes))
print("Naive Bayes Test F1 Score:", Metrics.f1_score(y_test.values, y_test_pred_bayes))

Naive Bayes Test Accuracy: 0.9497584541062802
Naive Bayes Test Precision: 0.8892307692307693
Naive Bayes Test Recall: 0.9475409836065574
Naive Bayes Test F1 Score: 0.9174603174603175


<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Performance Comparison</b><br>
Both models demonstrated excellent performance, with very similar overall results. However, there is a distinct trade-off in their behavior:
<br>
<table style="width:80%; margin: 10px auto; border-collapse: collapse;">
  <tr style="border-bottom: 1px solid #ddd;">
    <th style="text-align: left;">Metric</th>
    <th style="text-align: center;">Logistic Regression (Balanced)</th>
    <th style="text-align: center;">Multinomial Naive Bayes</th>
  </tr>
  <tr>
    <td><b>Recall</b></td>
    <td style="text-align: center; color: green;"><b>99.3%</b></td>
    <td style="text-align: center;">94.8%</td>
  </tr>
  <tr>
    <td><b>Precision</b></td>
    <td style="text-align: center;">84.9%</td>
    <td style="text-align: center; color: green;"><b>88.9%</b></td>
  </tr>
  <tr>
    <td><b>F1 Score</b></td>
    <td style="text-align: center;">0.915</td>
    <td style="text-align: center; color: green;"><b>0.917</b></td>
  </tr>
</table>
<br>
The <b>Logistic Regression</b> model achieved a near-perfect Recall (99.3%), making it ideal for security scenarios where missing a single spam email is unacceptable. However, this came at the cost of lower Precision (more false alarms).
<br>
<b>Naive Bayes</b> provided a slightly better overall balance (higher F1 and Precision), meaning it is less likely to accidentally classify a legitimate email as spam.
<br><br>

<b>2. Efficiency & Training Speed</b><br>
Despite the similar accuracy, <b>Naive Bayes</b> is the superior choice in terms of computational efficiency. It requires a single pass over the data to count frequencies, whereas Logistic Regression requires iterative optimization (Gradient Descent) to update weights over many epochs.
<br><br>

<b>3. Why does a "Naive" model work so well?</b><br>
It is surprising that a simple statistical model works so well despite the clearly false assumption that words are "independent" (e.g., ignoring that "free" and "money" belong together).
<br>
This works because spam classification relies on <b>strong independent signals</b>. The presence of specific keywords (<i>"winner", "click", "urgent"</i>) is often enough to trigger a classification, regardless of their grammatical relationship. Even if the probability estimates are not perfectly calibrated, the model correctly <b>ranks</b> the spam probability higher than the ham probability, resulting in the correct classification.
</p>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال دوم: استفاده از Logistic Regression و Naive Bayes <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">هدف این سوال، بررسی و مقایسه عملکرد مدل‌های Logistic Regression و Naive Bayes در تشخیص لینک‌های فیشینگ از لینک‌های قانونی (معتبر) است.
شما باید با استفاده از مجموعه‌داده‌ی ارائه‌شده (آدرس‌های اینترنتی خام)، یک سری ویژگی‌های عددی از این URLها استخراج نمایید و تأثیر انتخاب ویژگی‌ها و نرمال‌سازی داده‌ها را بر عملکرد مدل‌ها بررسی کنید.<br>نکته: در این سوال مجاز به استفاده از کتاب‌خانه‌های آماده هستید.


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: ویژگی‌های ساختاری</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- تعداد نقطه‌ها (nb_dots)
<br>
- تعداد اسلش‌ها (nb_slashes)
<br>
- تعداد خط ‌تیره‌ها (nb_hyphens)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی بیست درصد داده‌ها (داده‌های آزمون) مقایسه کنید
<br>
<br>
- آیا لازم است این ویژگی‌ها را نرمال‌سازی کنید؟ چرا؟ اگر پاسخ شما «بله» است، نوع نرمال‌سازی را توضیح دهید و آن را اجرا کنید.
<br>
- به نظر شما از بین نسخه‌های مختلف Naive Bayes (GaussianNB یا MultinomialNB) کدام برای داده‌های شما مناسب‌تر است؟ دلیل خود را بنویسید و با آن مدل آزمایش را انجام دهید.
</p>

<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Normalization Strategy</b><br>
Normalization is essential for this comparison, but the choice of method is constrained by the specific requirements of our models:
<ul>
    <li><b>Logistic Regression:</b> Uses Gradient Descent optimization. If features have vastly different scales (e.g., URL length vs. number of dots), the gradients will be distorted, leading to slow convergence or poor performance. Normalization is strictly required.</li>
    <li><b>Multinomial Naive Bayes:</b> Operates on counts and <b>cannot accept negative input values</b>.</li>
</ul>
<br>
<b>Conclusion:</b> While StandardScaler (Z-score) is often preferred for Logistic Regression, it produces negative values. Therefore, we must use <b>Min-Max Scaling</b> to map all features to the range <b>[0, 1]</b>. This ensures efficient convergence for Logistic Regression while satisfying the non-negative constraint of Naive Bayes.
<br><br>

<b>2. Choosing the Naive Bayes Variant</b><br>
We chose <b>Multinomial Naive Bayes</b> over Gaussian Naive Bayes for the following reasons:
<br>
The <b>Gaussian NB</b> model assumes features follow a continuous Normal (Bell Curve) distribution. However, our structural features (counts of dots, hyphens, slashes) are <b>discrete integers</b>, not continuous values. These counts often follow a skewed distribution rather than a symmetric bell curve.
<br>
<b>Multinomial NB</b> is explicitly designed to model discrete counts, making it the theoretically correct choice for feature sets based on frequency or occurrence counts.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

We will first load the dataset and take a quick look at it

In [39]:
url_dataset = pd.read_csv('./datasets/q2/urls.csv')
url_dataset.head()

,url,status
0,http://www.crestonwood.com/router.php,legitimate
1,http://shadetreetechnology.com/V4/validation/a...,phishing
2,https://support-appleld.com.secureupdate.duila...,phishing
3,http://rgipt.ac.in,legitimate
4,http://www.iracing.com/tracks/gateway-motorspo...,legitimate


Now let's extract features and change status to a binary value

In [40]:
status_mapping = {'legitimate': 0, 'phishing': 1}
url_dataset['status'] = url_dataset['status'].map(status_mapping)

In [41]:
url_dataset['nb_dots'] = url_dataset['url'].apply(lambda x: x.count('.'))
url_dataset['nb_slashes'] = url_dataset['url'].apply(lambda x: x.count('/'))
url_dataset['nb_hyphens'] = url_dataset['url'].apply(lambda x: x.count('-'))

url_dataset.head()

,url,status,nb_dots,nb_slashes,nb_hyphens
0,http://www.crestonwood.com/router.php,0,3,3,0
1,http://shadetreetechnology.com/V4/validation/a...,1,1,5,0
2,https://support-appleld.com.secureupdate.duila...,1,4,5,1
3,http://rgipt.ac.in,0,2,2,0
4,http://www.iracing.com/tracks/gateway-motorspo...,0,2,5,2


Now let's split the data

In [42]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    url_dataset.drop(columns=['url', 'status']),
    url_dataset['status'],
    test_size=0.2,
    random_state=random_state
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=random_state
)

print("Training set size:", X_train.shape)
print("Validation set size:", X_val.shape)
print("Test set size:", X_test.shape)

Training set size: (7315, 3)
Validation set size: (1829, 3)
Test set size: (2286, 3)


Now before training, it's best for the logistic regression models that the data is normalised

In [43]:
from sklearn.preprocessing import MinMaxScaler
normalizer_instance = MinMaxScaler()
X_train_normalized = normalizer_instance.fit_transform(X_train)
X_val_normalized = normalizer_instance.transform(X_val)
X_test_normalized = normalizer_instance.transform(X_test)

print(X_train_normalized)

[[0.04347826 0.03225806 0.        ]
 [0.13043478 0.16129032 0.        ]
 [0.13043478 0.06451613 0.02325581]
 ...
 [0.13043478 0.09677419 0.02325581]
 [0.08695652 0.09677419 0.04651163]
 [0.04347826 0.06451613 0.02325581]]


Now let's train the models, I will train both the Bayes models despite thinking that Multinomial is better to check the results.

In [44]:
from sklearn import linear_model, metrics
from sklearn.naive_bayes import MultinomialNB, GaussianNB

log_reg_model = linear_model.LogisticRegression()
log_reg_model.fit(X_train_normalized, y_train)

multinomial_naive_bayes_model = MultinomialNB()
multinomial_naive_bayes_model.fit(X_train_normalized, y_train)

gaussian_naive_bayes_model = GaussianNB()
gaussian_naive_bayes_model.fit(X_train_normalized, y_train)

def make_metrics(y_true, y_pred, pos_label=1):
    return {
        'accuracy': metrics.accuracy_score(y_true, y_pred),
        'precision': metrics.precision_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'recall': metrics.recall_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'f1': metrics.f1_score(y_true, y_pred, pos_label=pos_label, zero_division=0)
    }

y_train_pred_log = log_reg_model.predict(X_train_normalized)
y_val_pred_log   = log_reg_model.predict(X_val_normalized)
y_test_pred_log  = log_reg_model.predict(X_test_normalized)

y_train_pred_mnb = multinomial_naive_bayes_model.predict(X_train_normalized)
y_val_pred_mnb   = multinomial_naive_bayes_model.predict(X_val_normalized)
y_test_pred_mnb  = multinomial_naive_bayes_model.predict(X_test_normalized)

y_train_pred_gnb = gaussian_naive_bayes_model.predict(X_train_normalized)
y_val_pred_gnb   = gaussian_naive_bayes_model.predict(X_val_normalized)
y_test_pred_gnb  = gaussian_naive_bayes_model.predict(X_test_normalized)

metrics_train = {
    'LogisticRegression': make_metrics(y_train, y_train_pred_log),
    'MultinomialNB': make_metrics(y_train, y_train_pred_mnb),
    'GaussianNB': make_metrics(y_train, y_train_pred_gnb)
}

metrics_val = {
    'LogisticRegression': make_metrics(y_val, y_val_pred_log),
    'MultinomialNB': make_metrics(y_val, y_val_pred_mnb),
    'GaussianNB': make_metrics(y_val, y_val_pred_gnb)
}

metrics_test = {
    'LogisticRegression': make_metrics(y_test, y_test_pred_log),
    'MultinomialNB': make_metrics(y_test, y_test_pred_mnb),
    'GaussianNB': make_metrics(y_test, y_test_pred_gnb)
}

df_train_metrics_3structure = pd.DataFrame(metrics_train).T
df_val_metrics   = pd.DataFrame(metrics_val).T
df_test_metrics  = pd.DataFrame(metrics_test).T

df_train_metrics_3structure = df_train_metrics_3structure.round(4)
df_val_metrics_3structure   = df_val_metrics.round(4)
df_test_metrics_3structure  = df_test_metrics.round(4)

print("Train metrics")
display(df_train_metrics_3structure)
print("Validation metrics")
display(df_val_metrics)
print("Test metrics")
display(df_test_metrics)

structure_results = pd.concat(
    {
        'train': df_train_metrics_3structure,
        'val':   df_val_metrics_3structure,
        'test':  df_test_metrics_3structure
    },
    axis=1
)
display(structure_results)

Train metrics


,accuracy,precision,recall,f1
LogisticRegression,0.6809,0.7378,0.5639,0.6393
MultinomialNB,0.5580,0.5383,0.8309,0.6534
GaussianNB,0.6771,0.7764,0.4999,0.6082


Validation metrics


,accuracy,precision,recall,f1
LogisticRegression,0.691088,0.743448,0.587146,0.656117
MultinomialNB,0.552761,0.534722,0.838780,0.653096
GaussianNB,0.684527,0.775444,0.522876,0.624593


Test metrics


,accuracy,precision,recall,f1
LogisticRegression,0.679353,0.736029,0.547788,0.628108
MultinomialNB,0.528434,0.514840,0.798230,0.625954
GaussianNB,0.669729,0.770563,0.472566,0.585848


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6809    0.7378  0.5639  0.6393   0.6911    0.7434   
MultinomialNB        0.5580    0.5383  0.8309  0.6534   0.5528    0.5347   
GaussianNB           0.6771    0.7764  0.4999  0.6082   0.6845    0.7754   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5871  0.6561   0.6794    0.7360  0.5478  0.6281  
MultinomialNB       0.8388  0.6531   0.5284    0.5148  0.7982  0.6260  
GaussianNB          0.5229  0.6246   0.6697    0.7706  0.4726  0.5858

We should only use the test set once but since the last question wanted the result on the test set, I assumed here's the same.

Looking at the results we can see that all models performed quite similarly but the gaussian NB had a higher f1 score (which was because of the unusually high recall). The Logistic regression had the highers accuracy and precision but it's recall was noticeably lower that the Bayes models.

Since we are going to add features, I didn't try to improve the models at this stage.

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: ویژگی‌های آماری و محتوایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی زیر را از آدرس استخراج کنید:
<br>
- طول کل آدرس (length_url)
<br>
- نسبت تعداد ارقام (0–9) به کل طول آدرس (ratio_digits_url)
<br>
- طولانی‌ترین کلمه (کاراکترهای الفبایی) در URL که با علائم جداساز (نقطه، علامت‌سوال، اسلش و...) از هم جدا شده‌اند. (longest_words_raw)
<br>
سپس با این سه ویژگی مدل‌ها را روی هشتاد درصد داده‌ها (داده‌های آموزش) آموزش دهید و روی داده‌های آزمون (بیست درصد داده‌ها) مقایسه کنید.
<br>
- به نظر شما استفاده از Naive Bayes برای این ویژگی‌ها منطقی است؟ توضیح دهید. کدام نسخه مناسب‌تر است؟ 
</p>

<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Feature Analysis & Model Suitability</b><br>
In this section, we introduced features that fundamentally differ from the previous "count-based" features:
<ul>
    <li><b>ratio_digits_url:</b> This is a <b>continuous</b> value (e.g., 0.15, 0.42) rather than a discrete integer count.</li>
    <li><b>length_url:</b> While a count, this has a much larger magnitude than typical word frequencies.</li>
</ul>
<br>
<b>Multinomial Naive Bayes</b> is theoretically designed for integer counts. However, by applying Min-Max scaling (0–1 range), we can force it to work with these continuous features, though it technically violates the underlying assumption of the model.
<br>
<b>Gaussian Naive Bayes</b> assumes features follow a Normal distribution. Theoretically, it is better suited for continuous features like ratios.
<br><br>

<b>2. Results Analysis</b><br>
<table style="width:95%; margin: 10px auto; border-collapse: collapse; font-size: 0.9em;">
  <tr style="border-bottom: 1px solid #ddd; background-color: #f9f9f9;">
    <th style="text-align: left; padding: 8px;">Model (Test Set)</th>
    <th style="text-align: center;">Accuracy</th>
    <th style="text-align: center;">Precision</th>
    <th style="text-align: center;">Recall</th>
    <th style="text-align: center;">F1 Score</th>
  </tr>
  <tr>
    <td style="padding: 5px;"><b>Logistic Regression</b></td>
    <td style="text-align: center;"><b>65.0%</b></td>
    <td style="text-align: center;">72.6%</td>
    <td style="text-align: center;"><b>46.7%</b></td>
    <td style="text-align: center;"><b>0.57</b></td>
  </tr>
  <tr>
    <td style="padding: 5px;"><b>Multinomial NB</b></td>
    <td style="text-align: center;">62.9%</td>
    <td style="text-align: center;">68.9%</td>
    <td style="text-align: center;">45.4%</td>
    <td style="text-align: center;">0.55</td>
  </tr>
  <tr>
    <td style="padding: 5px;"><b>Gaussian NB</b></td>
    <td style="text-align: center;">64.0%</td>
    <td style="text-align: center;"><b>78.8%</b></td>
    <td style="text-align: center;">37.3%</td>
    <td style="text-align: center;">0.51</td>
  </tr>
</table>
<br>
The results tell a nuanced story:
<ul>
    <li><b>Logistic Regression</b> achieved the best overall balance (highest F1 and Accuracy). It effectively managed the trade-off between precision and recall.</li>
    <li><b>Multinomial NB</b> performed surprisingly well. Normalizing the continuous features to the [0, 1] range allowed it to function effectively, despite the theoretical mismatch.</li>
    <li><b>Gaussian NB</b> was the most "conservative" model. It achieved the <b>highest Precision (~79%)</b>, meaning when it flagged a URL as phishing, it was usually right. However, it had the <b>lowest Recall (~37%)</b>, meaning it missed the majority of actual phishing attacks.</li>
</ul>
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [45]:
url_dataset = pd.read_csv('./datasets/q2/urls.csv')
status_mapping = {'legitimate': 0, 'phishing': 1}
url_dataset['status'] = url_dataset['status'].map(status_mapping)

We will now add the new features.

In [46]:
url_dataset['length_url'] = url_dataset['url'].apply(lambda x: len(x))
url_dataset['ratio_digits_url'] = url_dataset['url'].apply(lambda x: sum(c.isdigit() for c in x) / len(x) if len(x) > 0 else 0)

import re

def get_longest_word_length(url):
    words = re.split(r'[^a-zA-Z]+', url)
    
    words = [w for w in words if w]
    if not words:
        return 0
    return max(len(w) for w in words)

url_dataset['longest_words_raw'] = url_dataset['url'].apply(get_longest_word_length)

url_dataset.head()

,url,status,length_url,ratio_digits_url,longest_words_raw
0,http://www.crestonwood.com/router.php,0,37,0.000000,11
1,http://shadetreetechnology.com/V4/validation/a...,1,77,0.220779,19
2,https://support-appleld.com.secureupdate.duila...,1,126,0.150794,13
3,http://rgipt.ac.in,0,18,0.000000,5
4,http://www.iracing.com/tracks/gateway-motorspo...,0,55,0.000000,11


Now we split the data just like before

In [47]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    url_dataset.drop(columns=['url', 'status']),
    url_dataset['status'],
    test_size=0.2,
    random_state=random_state
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=random_state
)

print("Training set size:", X_train.shape)
print("Validation set size:", X_val.shape)
print("Test set size:", X_test.shape)

Training set size: (7315, 3)
Validation set size: (1829, 3)
Test set size: (2286, 3)


And normalize it the same way

In [48]:
from sklearn.preprocessing import MinMaxScaler
normalizer_instance = MinMaxScaler()
X_train_normalized = normalizer_instance.fit_transform(X_train)
X_val_normalized = normalizer_instance.transform(X_val)
X_test_normalized = normalizer_instance.transform(X_test)

print(X_train_normalized)

[[0.0092081  0.         0.04545455]
 [0.04235727 0.         0.09848485]
 [0.07059546 0.21755013 0.06818182]
 ...
 [0.03192142 0.19426546 0.03787879]
 [0.03192142 0.         0.11363636]
 [0.02148557 0.         0.06060606]]


Finally train and compare the models the same as before.

In [49]:
from sklearn import linear_model, metrics
from sklearn.naive_bayes import MultinomialNB, GaussianNB

log_reg_model = linear_model.LogisticRegression()
log_reg_model.fit(X_train_normalized, y_train)

multinomial_naive_bayes_model = MultinomialNB()
multinomial_naive_bayes_model.fit(X_train_normalized, y_train)

gaussian_naive_bayes_model = GaussianNB()
gaussian_naive_bayes_model.fit(X_train_normalized, y_train)

def make_metrics(y_true, y_pred, pos_label=1):
    return {
        'accuracy': metrics.accuracy_score(y_true, y_pred),
        'precision': metrics.precision_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'recall': metrics.recall_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'f1': metrics.f1_score(y_true, y_pred, pos_label=pos_label, zero_division=0)
    }

y_train_pred_log = log_reg_model.predict(X_train_normalized)
y_val_pred_log   = log_reg_model.predict(X_val_normalized)
y_test_pred_log  = log_reg_model.predict(X_test_normalized)

y_train_pred_mnb = multinomial_naive_bayes_model.predict(X_train_normalized)
y_val_pred_mnb   = multinomial_naive_bayes_model.predict(X_val_normalized)
y_test_pred_mnb  = multinomial_naive_bayes_model.predict(X_test_normalized)

y_train_pred_gnb = gaussian_naive_bayes_model.predict(X_train_normalized)
y_val_pred_gnb   = gaussian_naive_bayes_model.predict(X_val_normalized)
y_test_pred_gnb  = gaussian_naive_bayes_model.predict(X_test_normalized)

metrics_train = {
    'LogisticRegression': make_metrics(y_train, y_train_pred_log),
    'MultinomialNB': make_metrics(y_train, y_train_pred_mnb),
    'GaussianNB': make_metrics(y_train, y_train_pred_gnb)
}

metrics_val = {
    'LogisticRegression': make_metrics(y_val, y_val_pred_log),
    'MultinomialNB': make_metrics(y_val, y_val_pred_mnb),
    'GaussianNB': make_metrics(y_val, y_val_pred_gnb)
}

metrics_test = {
    'LogisticRegression': make_metrics(y_test, y_test_pred_log),
    'MultinomialNB': make_metrics(y_test, y_test_pred_mnb),
    'GaussianNB': make_metrics(y_test, y_test_pred_gnb)
}

df_train_metrics_3statistic = pd.DataFrame(metrics_train).T
df_val_metrics_3statistic = pd.DataFrame(metrics_val).T
df_test_metrics_3statistic = pd.DataFrame(metrics_test).T

df_train_metrics_3statistic = df_train_metrics_3statistic.round(4)
df_val_metrics_3statistic   = df_val_metrics_3statistic.round(4)
df_test_metrics_3statistic  = df_test_metrics_3statistic.round(4)

# print("Train metrics")
# display(df_train_metrics_3statistic)
# print("Validation metrics")
# display(df_val_metrics_3statistic)
# print("Test metrics")
# display(df_test_metrics_3statistic)

statistic_results = pd.concat(
    {
        'train': df_train_metrics_3statistic,
        'val':   df_val_metrics_3statistic,
        'test':  df_test_metrics_3statistic
    },
    axis=1
)

print("Statistic-based Feature Results:")
display(statistic_results)

print("Structure-based Feature Results:")
display(structure_results)

Statistic-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6625    0.7306  0.5176  0.6059   0.6659    0.7351   
MultinomialNB        0.6316    0.6811  0.4985  0.5757   0.6326    0.6804   
GaussianNB           0.6484    0.7933  0.4039  0.5352   0.6424    0.7809   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5229  0.6111   0.6496    0.7263  0.4673  0.5687  
MultinomialNB       0.5054  0.5800   0.6290    0.6895  0.4540  0.5475  
GaussianNB          0.3998  0.5288   0.6404    0.7884  0.3726  0.5060

Structure-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6809    0.7378  0.5639  0.6393   0.6911    0.7434   
MultinomialNB        0.5580    0.5383  0.8309  0.6534   0.5528    0.5347   
GaussianNB           0.6771    0.7764  0.4999  0.6082   0.6845    0.7754   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5871  0.6561   0.6794    0.7360  0.5478  0.6281  
MultinomialNB       0.8388  0.6531   0.5284    0.5148  0.7982  0.6260  
GaussianNB          0.5229  0.6246   0.6697    0.7706  0.4726  0.5858

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: ویژگی‌های خلاقانه</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
سه ویژگی به انتخاب خودتان از آدرس‌ها استخراج نمایید و مشابه دو بخش قبلی، مدل‌ها را آموزش دهید.
برای هر ویژگی توضیح دهید چرا ممکن است در تشخیص فیشینگ مؤثر باشد؟
</p>

<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Rationale for Creative Features</b><br>
We introduced three binary features to capture specific "red flags" common in phishing URLs:
<ul>
    <li><b>has_ip:</b> Phishing sites often host content on compromised servers without purchasing a domain name, resulting in raw IP addresses in the URL (e.g., <i>http://192.168.1.1/...</i>). Legitimate public sites rarely do this.</li>
    <li><b>has_at_symbol (@):</b> This character is used in URLs to separate user credentials from the domain (<i>user:pass@domain.com</i>). Phishers abuse this to trick browsers and users (e.g., <i>google.com@malicious.com</i>), where the user sees "google.com" but is actually directed to "malicious.com".</li>
    <li><b>has_https:</b> While modern phishing sites often use free SSL certificates, the <i>absence</i> of HTTPS is a strong indicator of a low-security or illegitimate website, as nearly all modern legitimate services enforce encryption.</li>
</ul>
<br>

<b>2. Model Analysis</b><br>
An interesting observation from the results is that <b>Logistic Regression and Multinomial Naive Bayes produced identical metrics</b>.
<br>
This occurs because our features are purely <b>binary</b> (0 or 1). In this specific binary feature space, the "count-based" probability logic of Naive Bayes and the "weight-based" decision boundary of Logistic Regression often converge mathematically, leading to the exact same classification outputs.
</p>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
- جدولی (دیتافریم) شامل ۳ ویژگی استخراج‌شده + برچسب هدف (status)<br>
- جدول مقایسه‌ای شامل Accuracy, Precision, Recall, F1-score برای هر مدل
</p>

In [59]:
url_dataset = pd.read_csv('./datasets/q2/urls.csv')
status_mapping = {'legitimate': 0, 'phishing': 1}
url_dataset['status'] = url_dataset['status'].map(status_mapping)

We will now add the new features.

In [60]:
url_dataset['has_ip'] = url_dataset['url'].apply(lambda x: 1 if re.search(r'\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b', x) else 0)
url_dataset['has_at_symbol'] = url_dataset['url'].apply(lambda x: 1 if '@' in x else 0)
url_dataset['has_https'] = url_dataset['url'].apply(lambda x: 1 if 'https' in x else 0)

url_dataset.head()

,url,status,has_ip,has_at_symbol,has_https
0,http://www.crestonwood.com/router.php,0,0,0,0
1,http://shadetreetechnology.com/V4/validation/a...,1,0,0,0
2,https://support-appleld.com.secureupdate.duila...,1,0,0,1
3,http://rgipt.ac.in,0,0,0,0
4,http://www.iracing.com/tracks/gateway-motorspo...,0,0,0,0


Now we split the data just like before

In [61]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    url_dataset.drop(columns=['url', 'status']),
    url_dataset['status'],
    test_size=0.2,
    random_state=random_state
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=random_state
)

print("Training set size:", X_train.shape)
print("Validation set size:", X_val.shape)
print("Test set size:", X_test.shape)

Training set size: (7315, 3)
Validation set size: (1829, 3)
Test set size: (2286, 3)


And normalize it the same way

In [63]:
from sklearn.preprocessing import MinMaxScaler
normalizer_instance = MinMaxScaler()
X_train_normalized = normalizer_instance.fit_transform(X_train)
X_val_normalized = normalizer_instance.transform(X_val)
X_test_normalized = normalizer_instance.transform(X_test)

print(X_train_normalized)

[[0. 0. 1.]
 [0. 0. 0.]
 [0. 0. 1.]
 ...
 [0. 0. 1.]
 [0. 0. 0.]
 [0. 0. 0.]]


Finally train and compare the models the same as before.

In [64]:
from sklearn import linear_model, metrics
from sklearn.naive_bayes import MultinomialNB, GaussianNB

log_reg_model = linear_model.LogisticRegression()
log_reg_model.fit(X_train_normalized, y_train)

multinomial_naive_bayes_model = MultinomialNB()
multinomial_naive_bayes_model.fit(X_train_normalized, y_train)

gaussian_naive_bayes_model = GaussianNB()
gaussian_naive_bayes_model.fit(X_train_normalized, y_train)

def make_metrics(y_true, y_pred, pos_label=1):
    return {
        'accuracy': metrics.accuracy_score(y_true, y_pred),
        'precision': metrics.precision_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'recall': metrics.recall_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'f1': metrics.f1_score(y_true, y_pred, pos_label=pos_label, zero_division=0)
    }

y_train_pred_log = log_reg_model.predict(X_train_normalized)
y_val_pred_log   = log_reg_model.predict(X_val_normalized)
y_test_pred_log  = log_reg_model.predict(X_test_normalized)

y_train_pred_mnb = multinomial_naive_bayes_model.predict(X_train_normalized)
y_val_pred_mnb   = multinomial_naive_bayes_model.predict(X_val_normalized)
y_test_pred_mnb  = multinomial_naive_bayes_model.predict(X_test_normalized)

y_train_pred_gnb = gaussian_naive_bayes_model.predict(X_train_normalized)
y_val_pred_gnb   = gaussian_naive_bayes_model.predict(X_val_normalized)
y_test_pred_gnb  = gaussian_naive_bayes_model.predict(X_test_normalized)

metrics_train = {
    'LogisticRegression': make_metrics(y_train, y_train_pred_log),
    'MultinomialNB': make_metrics(y_train, y_train_pred_mnb),
    'GaussianNB': make_metrics(y_train, y_train_pred_gnb)
}

metrics_val = {
    'LogisticRegression': make_metrics(y_val, y_val_pred_log),
    'MultinomialNB': make_metrics(y_val, y_val_pred_mnb),
    'GaussianNB': make_metrics(y_val, y_val_pred_gnb)
}

metrics_test = {
    'LogisticRegression': make_metrics(y_test, y_test_pred_log),
    'MultinomialNB': make_metrics(y_test, y_test_pred_mnb),
    'GaussianNB': make_metrics(y_test, y_test_pred_gnb)
}

df_train_metrics_3creative = pd.DataFrame(metrics_train).T
df_val_metrics_3creative = pd.DataFrame(metrics_val).T
df_test_metrics_3creative = pd.DataFrame(metrics_test).T

df_train_metrics_3creative = df_train_metrics_3creative.round(4)
df_val_metrics_3creative   = df_val_metrics_3creative.round(4)
df_test_metrics_3creative  = df_test_metrics_3creative.round(4)

# print("Train metrics")
# display(df_train_metrics_3creative)
# print("Validation metrics")
# display(df_val_metrics_3creative)
# print("Test metrics")
# display(df_test_metrics_3creative)

creative_results = pd.concat(
    {
        'train': df_train_metrics_3creative,
        'val':   df_val_metrics_3creative,
        'test':  df_test_metrics_3creative
    },
    axis=1
)
print("Creative Feature Results:")
display(creative_results)

print("Statistic-based Feature Results:")
display(statistic_results)

print("Structure-based Feature Results:")
display(structure_results)

Creative Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.5645    0.5535  0.6790  0.6098   0.5599    0.5516   
MultinomialNB        0.5645    0.5535  0.6790  0.6098   0.5599    0.5516   
GaussianNB           0.5292    0.9912  0.0614  0.1156   0.5265    1.0000   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.6580  0.6001   0.5494    0.5346  0.6832  0.5998  
MultinomialNB       0.6580  0.6001   0.5494    0.5346  0.6832  0.5998  
GaussianNB          0.0566  0.1072   0.5306    1.0000  0.0504  0.0960

Statistic-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6625    0.7306  0.5176  0.6059   0.6659    0.7351   
MultinomialNB        0.6316    0.6811  0.4985  0.5757   0.6326    0.6804   
GaussianNB           0.6484    0.7933  0.4039  0.5352   0.6424    0.7809   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5229  0.6111   0.6496    0.7263  0.4673  0.5687  
MultinomialNB       0.5054  0.5800   0.6290    0.6895  0.4540  0.5475  
GaussianNB          0.3998  0.5288   0.6404    0.7884  0.3726  0.5060

Structure-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6809    0.7378  0.5639  0.6393   0.6911    0.7434   
MultinomialNB        0.5580    0.5383  0.8309  0.6534   0.5528    0.5347   
GaussianNB           0.6771    0.7764  0.4999  0.6082   0.6845    0.7754   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5871  0.6561   0.6794    0.7360  0.5478  0.6281  
MultinomialNB       0.8388  0.6531   0.5284    0.5148  0.7982  0.6260  
GaussianNB          0.5229  0.6246   0.6697    0.7706  0.4726  0.5858

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: ترکیب همه ویژگی‌ها</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
تمام ۹ ویژگی (۳ ساختاری + ۳ آماری + ۳ خلاقانه) را با هم ترکیب کنید.
سپس هر دو مدل را دوباره آموزش دهید و نتایج را ارزیابی کنید.
<br>
آیا ترکیب همه ویژگی‌ها باعث بهبود عملکرد مدل می‌شود؟
<br>
اگر خیر، به نظر شما دلیل آن چیست؟ (مثلاً تداخل بین ویژگی‌ها یا همبستگی بالا و...)
</p>


<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. The "Super Model" Hypothesis</b><br>
We initially hypothesized that combining all features would improve performance by providing a holistic view (Structure + Content + Red Flags).
<br>
The results confirm this for <b>Logistic Regression</b>, which achieved its highest accuracy yet (~71%). This demonstrates that when a model has the capacity to weigh inputs (assigning high coefficients to strong features and low coefficients to weak ones), adding relevant information generally improves performance.
<br><br>

<b>2. The Naive Bayes Struggle</b><br>
However, the Naive Bayes models did not benefit as universally. This highlights a critical limitation regarding <b>distribution assumptions</b>:
<ul>
    <li><b>Gaussian NB Performance:</b> This model performed poorly (F1 Score ~0.30). This is because we introduced <b>binary features</b> (0 and 1) into the dataset. Gaussian NB assumes data follows a Normal Distribution (Bell Curve). Binary data is the exact opposite of a Bell Curve! This severe mismatch confused the model's probability calculations.</li>
    <li><b>Feature Correlation:</b> Naive Bayes assumes all features are <b>independent</b>. By combining "Length of URL" (Statistical) with "Number of Dots" (Structural), we likely introduced redundant, correlated signals. This violation of the independence assumption can make the model overconfident in the wrong direction, hurting performance compared to the discriminative Logistic Regression.</li>
</ul>
</p>

In [65]:
url_dataset = pd.read_csv('./datasets/q2/urls.csv')
status_mapping = {'legitimate': 0, 'phishing': 1}
url_dataset['status'] = url_dataset['status'].map(status_mapping)

In this step we will add the features of all previous sections!

In [66]:
url_dataset['nb_dots'] = url_dataset['url'].apply(lambda x: x.count('.'))
url_dataset['nb_slashes'] = url_dataset['url'].apply(lambda x: x.count('/'))
url_dataset['nb_hyphens'] = url_dataset['url'].apply(lambda x: x.count('-'))

url_dataset['length_url'] = url_dataset['url'].apply(lambda x: len(x))
url_dataset['ratio_digits_url'] = url_dataset['url'].apply(lambda x: sum(c.isdigit() for c in x) / len(x) if len(x) > 0 else 0)

import re

def get_longest_word_length(url):
    words = re.split(r'[^a-zA-Z]+', url)
    
    words = [w for w in words if w]
    if not words:
        return 0
    return max(len(w) for w in words)

url_dataset['longest_words_raw'] = url_dataset['url'].apply(get_longest_word_length)

url_dataset['has_ip'] = url_dataset['url'].apply(lambda x: 1 if re.search(r'\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b', x) else 0)
url_dataset['has_at_symbol'] = url_dataset['url'].apply(lambda x: 1 if '@' in x else 0)
url_dataset['has_https'] = url_dataset['url'].apply(lambda x: 1 if 'https' in x else 0)

url_dataset.head()

,url,status,nb_dots,nb_slashes,nb_hyphens,length_url,ratio_digits_url,longest_words_raw,has_ip,has_at_symbol,has_https
0,http://www.crestonwood.com/router.php,0,3,3,0,37,0.000000,11,0,0,0
1,http://shadetreetechnology.com/V4/validation/a...,1,1,5,0,77,0.220779,19,0,0,0
2,https://support-appleld.com.secureupdate.duila...,1,4,5,1,126,0.150794,13,0,0,1
3,http://rgipt.ac.in,0,2,2,0,18,0.000000,5,0,0,0
4,http://www.iracing.com/tracks/gateway-motorspo...,0,2,5,2,55,0.000000,11,0,0,0


We then split the data just like before

In [67]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    url_dataset.drop(columns=['url', 'status']),
    url_dataset['status'],
    test_size=0.2,
    random_state=random_state
)

X_train, X_val, y_train, y_val = train_test_split(
    X_train,
    y_train,
    test_size=0.2,
    random_state=random_state
)

print("Training set size:", X_train.shape)
print("Validation set size:", X_val.shape)
print("Test set size:", X_test.shape)

Training set size: (7315, 9)
Validation set size: (1829, 9)
Test set size: (2286, 9)


And normalize it

In [68]:
from sklearn.preprocessing import MinMaxScaler
normalizer_instance = MinMaxScaler()
X_train_normalized = normalizer_instance.fit_transform(X_train)
X_val_normalized = normalizer_instance.transform(X_val)
X_test_normalized = normalizer_instance.transform(X_test)

print(X_train_normalized)

[[0.04347826 0.03225806 0.         ... 0.         0.         1.        ]
 [0.13043478 0.16129032 0.         ... 0.         0.         0.        ]
 [0.13043478 0.06451613 0.02325581 ... 0.         0.         1.        ]
 ...
 [0.13043478 0.09677419 0.02325581 ... 0.         0.         1.        ]
 [0.08695652 0.09677419 0.04651163 ... 0.         0.         0.        ]
 [0.04347826 0.06451613 0.02325581 ... 0.         0.         0.        ]]


Then train and compare the models

In [72]:
from sklearn import linear_model, metrics
from sklearn.naive_bayes import MultinomialNB, GaussianNB

log_reg_model = linear_model.LogisticRegression()
log_reg_model.fit(X_train_normalized, y_train)

multinomial_naive_bayes_model = MultinomialNB()
multinomial_naive_bayes_model.fit(X_train_normalized, y_train)

gaussian_naive_bayes_model = GaussianNB()
gaussian_naive_bayes_model.fit(X_train_normalized, y_train)

def make_metrics(y_true, y_pred, pos_label=1):
    return {
        'accuracy': metrics.accuracy_score(y_true, y_pred),
        'precision': metrics.precision_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'recall': metrics.recall_score(y_true, y_pred, pos_label=pos_label, zero_division=0),
        'f1': metrics.f1_score(y_true, y_pred, pos_label=pos_label, zero_division=0)
    }

y_train_pred_log = log_reg_model.predict(X_train_normalized)
y_val_pred_log   = log_reg_model.predict(X_val_normalized)
y_test_pred_log  = log_reg_model.predict(X_test_normalized)

y_train_pred_mnb = multinomial_naive_bayes_model.predict(X_train_normalized)
y_val_pred_mnb   = multinomial_naive_bayes_model.predict(X_val_normalized)
y_test_pred_mnb  = multinomial_naive_bayes_model.predict(X_test_normalized)

y_train_pred_gnb = gaussian_naive_bayes_model.predict(X_train_normalized)
y_val_pred_gnb   = gaussian_naive_bayes_model.predict(X_val_normalized)
y_test_pred_gnb  = gaussian_naive_bayes_model.predict(X_test_normalized)

metrics_train = {
    'LogisticRegression': make_metrics(y_train, y_train_pred_log),
    'MultinomialNB': make_metrics(y_train, y_train_pred_mnb),
    'GaussianNB': make_metrics(y_train, y_train_pred_gnb)
}

metrics_val = {
    'LogisticRegression': make_metrics(y_val, y_val_pred_log),
    'MultinomialNB': make_metrics(y_val, y_val_pred_mnb),
    'GaussianNB': make_metrics(y_val, y_val_pred_gnb)
}

metrics_test = {
    'LogisticRegression': make_metrics(y_test, y_test_pred_log),
    'MultinomialNB': make_metrics(y_test, y_test_pred_mnb),
    'GaussianNB': make_metrics(y_test, y_test_pred_gnb)
}

df_train_metrics_all_features = pd.DataFrame(metrics_train).T
df_val_metrics_all_features = pd.DataFrame(metrics_val).T
df_test_metrics_all_features = pd.DataFrame(metrics_test).T

df_train_metrics_all_features = df_train_metrics_all_features.round(4)
df_val_metrics_all_features   = df_val_metrics_all_features.round(4)
df_test_metrics_all_features  = df_test_metrics_all_features.round(4)

# print("Train metrics")
# display(df_train_metrics_all_features)
# print("Validation metrics")
# display(df_val_metrics_all_features)
# print("Test metrics")
# display(df_test_metrics_all_features)

all_features_results = pd.concat(
    {
        'train': df_train_metrics_all_features,
        'val':   df_val_metrics_all_features,
        'test':  df_test_metrics_all_features
    },
    axis=1
)
print("All Features Results:")
display(all_features_results)

print("Creative Feature Results:")
display(creative_results)

print("Statistic-based Feature Results:")
display(statistic_results)

print("Structure-based Feature Results:")
display(structure_results)

All Features Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.7252    0.7806  0.6286  0.6964   0.7201    0.7707   
MultinomialNB        0.6008    0.5836  0.7112  0.6411   0.5888    0.5760   
GaussianNB           0.6052    0.9654  0.2203  0.3588   0.6003    0.9561   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.6296  0.6930   0.7082    0.7714  0.5823  0.6636  
MultinomialNB       0.6852  0.6259   0.5722    0.5534  0.6965  0.6168  
GaussianNB          0.2135  0.3491   0.5892    0.9569  0.1770  0.2987

Creative Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.5645    0.5535  0.6790  0.6098   0.5599    0.5516   
MultinomialNB        0.5645    0.5535  0.6790  0.6098   0.5599    0.5516   
GaussianNB           0.5292    0.9912  0.0614  0.1156   0.5265    1.0000   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.6580  0.6001   0.5494    0.5346  0.6832  0.5998  
MultinomialNB       0.6580  0.6001   0.5494    0.5346  0.6832  0.5998  
GaussianNB          0.0566  0.1072   0.5306    1.0000  0.0504  0.0960

Statistic-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6625    0.7306  0.5176  0.6059   0.6659    0.7351   
MultinomialNB        0.6316    0.6811  0.4985  0.5757   0.6326    0.6804   
GaussianNB           0.6484    0.7933  0.4039  0.5352   0.6424    0.7809   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5229  0.6111   0.6496    0.7263  0.4673  0.5687  
MultinomialNB       0.5054  0.5800   0.6290    0.6895  0.4540  0.5475  
GaussianNB          0.3998  0.5288   0.6404    0.7884  0.3726  0.5060

Structure-based Feature Results:


train                                val            \
                   accuracy precision  recall      f1 accuracy precision   
LogisticRegression   0.6809    0.7378  0.5639  0.6393   0.6911    0.7434   
MultinomialNB        0.5580    0.5383  0.8309  0.6534   0.5528    0.5347   
GaussianNB           0.6771    0.7764  0.4999  0.6082   0.6845    0.7754   

                                       test                            
                    recall      f1 accuracy precision  recall      f1  
LogisticRegression  0.5871  0.6561   0.6794    0.7360  0.5478  0.6281  
MultinomialNB       0.8388  0.6531   0.5284    0.5148  0.7982  0.6260  
GaussianNB          0.5229  0.6246   0.6697    0.7706  0.4726  0.5858

## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش پنجم: جمع‌بندی نهایی</div>


<p dir="rtl" style="line-height: 1.8; text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نتیجه‌گیری نهایی خودتان از این سوال را ارائه دهید. برای مثال باید "حداقل" به سوال‌های زیر پاسخ دهید:
<br>
- کدام مدل برای این نوع داده بهتر عمل کرده است؟ چرا؟
<br>
- کدام نوع ویژگی بیشترین تأثیر را داشته است؟ کدام نوع کمترین؟ چرا؟
<br>
- نرمال‌سازی نیاز بوده است؟ اگر بله، برای هر دو مدل؟ چرا؟
<br>
- غیر از استخراح ویژگی از خود آدرس اینترنتی، چه رویکرد دیگری می‌توان اتخاذ نمود؟
</p>


<p style="direction: ltr; text-align: left; background:#fffbe6; font-family: sans-serif; border:1px dashed #f0ad4e; padding:15px; border-radius:8px; color:#111; line-height: 1.6;">
✍️ <b>Answer:</b><br><br>

<b>1. Best Performing Model</b><br>
<b>Logistic Regression</b> was the superior model for this task, achieving the highest accuracy (>70%) when all features were combined.
<br>
<i>Reasoning:</i> Logistic Regression is a discriminative model that learns optimal weights for each feature individually. This flexibility allows it to effectively handle a heterogeneous mix of feature types (binary flags, continuous ratios, and large integer counts) simultaneously.
<br>
In contrast, the Naive Bayes models struggled with this mixed dataset. <b>Gaussian NB</b> failed significantly when binary features were introduced because binary data violates the Normal Distribution assumption. <b>Multinomial NB</b> was constrained by its requirement for integer counts, making it less ideal for the continuous ratio features.
<br><br>

<b>2. Feature Impact Analysis</b><br>
<ul>
    <li><b>Most Impactful (Structural & Statistical):</b> Features like <i>URL length</i>, <i>dot counts</i>, and <i>digit ratios</i> provided the strongest baseline performance (~66%). Phishing URLs often exhibit distinct structural anomalies (e.g., excessively long subdomains) that are present in the majority of phishing samples, providing a consistent signal for the model.</li>
    <li><b>Least Impactful (Creative/Binary):</b> Features like <i>has_ip</i> or <i>has_at_symbol</i> yielded the lowest individual accuracy (~55%). While these are strong indicators of phishing when they appear, they are <b>sparse</b> features—meaning they are absent in the vast majority of samples. Their absence provides little information to distinguish legitimate sites from sophisticated phishing sites.</li>
</ul>
<br>

<b>3. The Role of Normalization</b><br>
Normalization was strictly necessary, but for different reasons for each algorithm:
<ul>
    <li><b>For Logistic Regression:</b> It prevents features with large magnitudes (e.g., <i>length_url</i> ≈ 100) from dominating features with small magnitudes (e.g., <i>ratio_digits</i> < 1) during gradient descent optimization.</li>
    <li><b>For Multinomial NB:</b> We utilized <b>Min-Max Scaling</b> specifically because this algorithm cannot mathematically handle negative inputs. Standard scaling (Z-score) would have produced negative values, causing the implementation to fail.</li>
</ul>
<br>

<b>4. Alternative Approaches</b><br>
Beyond manual feature extraction from the URL string, we could employ more advanced techniques:
<ul>
    <li><b>Lexical Embeddings (NLP):</b> Instead of counting specific characters, we could treat the URL as a character sequence and use <b>n-grams</b> or <b>Character-level CNNs</b> to automatically learn complex patterns.</li>
    <li><b>External Context (WHOIS & DNS):</b> Phishing domains are often registered only hours before an attack. Querying the "Domain Age" via WHOIS data would likely be a highly predictive feature.</li>
    <li><b>Visual Analysis:</b> We could use a headless browser to render the webpage and use Computer Vision to check if the visual layout mimics a known brand (e.g., a fake Google login page) hosted on a strange URL.</li>
</ul>
</p>

# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>



<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA{n}_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
